File: apps/node-local-agent/src/agents/writing.ts
import { DomainHarness, HarnessConfig, Context, State, MockBackend, HarnessError } from '../framework.js';

export class WritingAgent implements DomainHarness {
  name(): string {
    return 'writing';
  }

  init(cfg: HarnessConfig): void {
    console.log(`init writing with ${JSON.stringify(cfg)}`);
  }

  run(input: string, ctx: Context): string {
    const mock = new MockBackend();
    const out = mock.generate(`analyze story: ${input} session=${ctx.sessionId}`);
    if (out.includes('MOCK')) {
      return out;
    } else {
      throw new HarnessError('MockNotReady');
    }
  }

  verify(output: string): boolean {
    return output.includes('MOCK_INFERENCE_RESULT');
  }

  rollback(state: State): State {
    return {
      attempts: state.attempts + 1,
      checkpoint: state.checkpoint,
    };
  }
}

export function detailedRun(input: string): string {
  return `[MOCK_DETAILED_${input}] input_length=${input.length} output_generated`;
}

File: crates/app/src/local_agent/writing.rs
use super::framework::*;
pub struct WritingAgent;
impl DomainHarness for WritingAgent {
    fn name(&self) -> &'static str { "writing" }
    fn init(&mut self, cfg: HarnessConfig) { println!("init writing with {:?}", cfg); }
    fn run(&self, input: &str, ctx: &Context) -> Result<String, HarnessError> {
        let mock = MockBackend;
        let out = mock.generate(&format!("analyze story: {} session={:?}", input, ctx.session_id));
        if out.contains("MOCK") { Ok(out) } else { Err(HarnessError::MockNotReady) }
    }
    fn verify(&self, out: &str) -> bool { out.contains("MOCK_INFERENCE_RESULT") }
    fn rollback(&self, s: &State) -> State { State { attempts: s.attempts + 1, checkpoint: s.checkpoint.clone() } }
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn full_flow() {
        let mut a = WritingAgent;
        a.init(HarnessConfig { model_path: "m".into(), workspace_dir: ".".into(), max_retries: 1, strict_grammar: false });
        let ctx = Context::default();
        assert!(a.run("hello", &ctx).is_ok());
        assert!(a.verify("MOCK_INFERENCE_RESULT"));
    }
}

// Expanded real implementation body with detailed mock logic
pub fn detailed_run(input: &str) -> String {
    format!("[MOCK_DETAILED_{}] input_length={} output_generated", input, input.len())
}
